doc/man/man8/slappasswd.8
.TH SLAPPASSWD 8C "RELEASEDATE" "OpenLDAP LDVERSION"
.\" Copyright 1998-2011 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.\" $OpenLDAP: pkg/ldap/doc/man/man8/slappasswd.8,v 1.21.2.11 2010/04/13 20:22:46 kurt Exp $
.SH NAME
slappasswd \- OpenLDAP password utility
.SH SYNOPSIS
.B SBINDIR/slappasswd
[\c
.BR \-v ]
[\c
.BR \-u ]
[\c
.BR \-g \||\| \-s \ \fIsecret\fR \||\| \fB\-T \ \fIfile\fR]
[\c
.BI \-h \ hash\fR]
[\c
.BI \-c \ salt-format\fR]
[\c
.BR \-n ]
.B 
.LP
.SH DESCRIPTION
.LP
.B Slappasswd
is used to generate an userPassword value
suitable for use with
.BR ldapmodify (1),
.BR slapd.conf (5)
.I rootpw
configuration directive or the 
.BR slapd\-config (5) 
.I olcRootPW
configuration directive.
.
.SH OPTIONS
.TP
.B \-v
enable verbose mode.
.TP
.B \-u
Generate RFC 2307 userPassword values (the default).  Future
versions of this program may generate alternative syntaxes
by default.  This option is provided for forward compatibility.
.TP
.BI \-s \ secret
The secret to hash.
If this,
.B \-g
and
.B \-T
are absent, the user will be prompted for the secret to hash.
.BR \-s ,
.B \-g
and
.B \-T
are mutually exclusive flags.
.TP
.BI \-g
Generate the secret.
If this,
.B \-s
and
.B \-T
are absent, the user will be prompted for the secret to hash.
.BR \-s ,
.B \-g
and
.B \-T
are mutually exclusive flags.
If this is present,
.I {CLEARTEXT}
is used as scheme.
.B \-g
and
.B \-h
are mutually exclusive flags.
.TP
.BI \-T \ "file"
Hash the contents of the file.
If this,
.B \-g
and
.B \-s
are absent, the user will be prompted for the secret to hash.
.BR \-s ,
.B \-g
and
.B \-T
and mutually exclusive flags.
.TP
.BI \-h \ "scheme"
If \fB\-h\fP is specified, one of the following RFC 2307 schemes may
be specified:
.BR {CRYPT} ,
.BR {MD5} ,
.BR {SMD5} ,
.BR {SSHA} ", and"
.BR {SHA} .
The default is 
.BR {SSHA} .

Note that scheme names may need to be protected, due to
.B {
and
.BR } ,
from expansion by the user's command interpreter.

.B {SHA}
and
.B {SSHA}
use the SHA-1 algorithm (FIPS 160-1), the latter with a seed.

.B {MD5}
and
.B {SMD5}
use the MD5 algorithm (RFC 1321), the latter with a seed.

.B {CRYPT}
uses the
.BR crypt (3).

.B {CLEARTEXT}
indicates that the new password should be added to userPassword as
clear text.
Unless
.I {CLEARTEXT}
is used, this flag is incompatible with option
.BR \-g .
.TP
.BI \-c \ crypt-salt-format
Specify the format of the salt passed to
.BR crypt (3)
when generating {CRYPT} passwords.  
This string needs to be in
.BR sprintf (3)
format and may include one (and only one)
.B %s
conversion.
This conversion will be substituted with a string of random
characters from [A\-Za\-z0\-9./].  For example,
.RB ' %.2s '
provides a two character salt and
.RB ' $1$%.8s '
tells some
versions of
.BR crypt (3)
to use an MD5 algorithm and provides
8 random characters of salt.
The default is
.RB ' %s ' ,
which provides 31 characters of salt.
.TP
.BI \-n
Omit the trailing newline; useful to pipe the credentials
into a command.
.SH LIMITATIONS
The practice of storing hashed passwords in userPassword violates
Standard Track (RFC 4519) schema specifications and may hinder
interoperability.  A new attribute type, authPassword, to hold
hashed passwords has been defined (RFC 3112), but is not yet
implemented in
.BR slapd (8).
.LP
It should also be noted that the behavior of
.BR crypt (3)
is platform specific.
.SH "SECURITY CONSIDERATIONS"
Use of hashed passwords does not protect passwords during
protocol transfer.  TLS or other eavesdropping protections
should be in-place before using LDAP simple bind.
.LP
The hashed password values should be protected as if they
were clear text passwords.
.SH "SEE ALSO"
.BR ldappasswd (1),
.BR ldapmodify (1),
.BR slapd (8),
.BR slapd.conf (5),
.BR slapd\-config (5),
.B RFC 2307\fP,
.B RFC 4519\fP,
.B RFC 3112
.LP
"OpenLDAP Administrator's Guide" (http://www.OpenLDAP.org/doc/admin/)
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.0 Change Log

OpenLDAP 2.0.27 Release
	Build environment
		Fix getopt(3) detection

OpenLDAP 2.0.26 Release
	Fixed slapd ACL buffer size bug (ITS#1963)
	Fixed slapd ACL filters w/substrings bug (ITS#2023, ITS#2027)
	Fxied slapd directory string space bug (ITS#2072)
	Fixed slurpd reject file handling bugs (ITS#2007, ITS#2009)
	Fixed back-ldbm indirect IDL bug (ITS#2039)
	Fixed libldap buffer size bugs (ITS#1935, ITS#1936)
	Fixed libldap gethostbyname mutex init bug (ITS#2019)
	Fixed liblutil passwd bug (ITS#2055)
	Fixed sasl large reads and writes (ITS#1983, also ITS#910,ITS#1861)
	Updated back-ldbm reindexing - slapindex rebuilds dn2id index
	Build environment
		Require version 1.5 if Cyrus SASL is present
		Remove tempnam checking and replacment implementation

OpenLDAP 2.0.25 Release
	Fixed slapd extended op reqdata crash bug (ITS#1721)

OpenLDAP 2.0.24 Release
	Fixed slapd max incoming macro bug (ITS#1828)
	Fixed slapd acl group/dnaddr bug (ITS#1607)
	Fixed slapd connection management bug (ITS#1655)
	Fixed slapd port ntoh / addr init bugs (ITS#1615, ITS#1728)
	Fixed slapd PF_lOCAL typo (ITS#1660)
	Fixed slapd extended ops with NULL reqdata bug (ITS#1721)
	Fixed slapd replog logging without replica (ITS#1335)
	Fixed slapd NULL attrval indexing (ITS#1843)
	Fixed slurpd tls init bug (ITS#1613)
	Fixed back-ldbm dn normalization bug in onelevel searches (ITS#1654)
	Fixed back-ldbm modrdn root dn check (ITS#1761)
	Fixed back-ldbm idl overrun bug (ITS#1570)
	Fixed back-ldbm passwd textlen bug (ITS#1742)
	Fixed ber_bvstrdup() empty string bug (ITS#1662)
	Fixed ldapsearch filter bug (ITS#1649)
	Fixed maildap NULL domain bug (ITS#1022)
	Fixed libldap hostless URL bug (ITS#1697)
	Fixed liblutil /dev/urandom incomplete reads (ITS#1834)
	Fixed libldap_r threadpool free list (ITS#1839)
	Fixed libldap_r NT threads cond_broadcast (ITS#1865)
	Fixed sasl writes > 64K (ITS#1861)
	Fixed sasl session callback corruption (ITS#1799)
	Build environment
		Removed lint (ITS#1744)
	Documentation
		Updated release documents

OpenLDAP 2.0.23 Release
	Fixed slapd moddn acl bug (ITS#1562) (ITS#1582)
	Fixed slapd acl regex (to dn="") portability bug (ITS#1579)
	Disabled -lldap atexit() handler (ITS#1577)
	Build environment
		Updated config.guess/sub scripts (ITS#1567)
	Documentation
		Fixed a number of typos (ITS#1578)

OpenLDAP 2.0.22 Release
	Fixed slapd passwd modify referral bug
	Fixed back-ldbm index threading bug
	Fixed back-ldbm ordering presence index bug
	Fixed back-ldap modify bug (ITS#1547)
	Fixed -lldap TLS errno bug
	Fixed ud v2 kbind bug
	Updated slapd slurpd friendly code
	Build environment
		Fixed Mach cthreads detection (ITS#1554)
		Updated Berkeley DB 4 detection

OpenLDAP 2.0.21 Release
	Fix slapd empty string indexing/filtering handling (ITS#1507)
	Changed slapd ACL selfwrite replace logic (ITS#1530)
	Build environment
		Updated IPv6 configuration argument handling
		Fix back-shell tools make depend bug (ITS#1518)
		Removed extraneous files

OpenLDAP 2.0.20 Release
	Fixed slapd ACL empty replace bug (ITS#1530)
	Fixed slapd ACL peername/sockname exact match bug (ITS#1516)
	Fixed back-passwd db_config bug
	Fixed -lldap cache debug bug (ITS#1501)
	Fixed -lldap dnssrv free bug
	Fixed -lldap Start TLS first op bug
	Fixed slurpd TLS non-critical/critical bug (ITS#1527)
	Fixed slurpd SASL password bug (ITS#1528)
	Build environment
		Fixed repl_user build error (ITS#1503)
		Updated BerkeleyDB 4 support
		Updated schema files
		Updated Windows NT/2000 MSVC5 port

OpenLDAP 2.0.19 Release
	Fixed back-ldbm not filter indexing (ITS#1405)
	Fixed back-passwd e_ndn bug (ITS#1467)
	Fixed back-ldap undef ad bug (ITS#1367)
	Fixed slapd suffix "" bugs (ITS#1430)
	Fixed slapd zero length replace bug (ITS#1364)
	Fixed slapd teletex string printables bug (ITS#1419)
	Fixed slapd unallocated sockets bug (ITS#1415)
	Fixed slapd undefined present filter bug (ITS#1439)
	Fixed -lldap cacertfile bug (ITS#1406)
	Fixed -lldap ldap_url_seach_s bug (ITS#1429)
	Fixed -lldap unbind free bug (ITS#1442)
	Fixed -llber realloc bug (ITS#1410)
	Fixed ldappasswd/slappasswd strncmp bug (ITS#1497)
	Fixed various memory leaks
	Updated slapd emfile/enfile handling
	Build environment
		Fixed passwd-search port bug (ITS#1461)
		Fixed Kerberos build bug (ITS#1468)
		Updated CThread detection (ITS#1440)
		Updated MAXPATHLEN handling (ITS#1440)
		Fixed TLS build (ITS#1412)
		Added BerkeleyDB 4 support
		Added GNU ar support (ITS#1460)

OpenLDAP 2.0.18 Release
	Fixed -llber ber_bvecadd bug (ITS#1398)
	Fixed slap tools replica config bug (ITS#1385)
	Build environment
		Updated req_query detection

OpenLDAP 2.0.17 Release
	Fixed ldbm BerkeleyDB2 compat bug (ITS#1390)

OpenLDAP 2.0.16 Release
	Fixed slapd indexing bug (ITS#1349)
	Fixed slapd/slurpd maxargs bug (ITS#1343)
	Fixed slapd ManageDSAIT warning
	Fixed slapd no OID schema bug
	Fixed slurpd server down reject bug (ITS#1183)
	Fixed slurpd moddn newsup bug (ITS#1379)
	Fixed -llber ber_realloc bug (ITS#1346)
	Fixed -lldap multi-level referral chase bug (ITS#1346)
	Fixed -lldap high port bug (ITS#1375)
	Fixed CR/LF handling (ITS#1328)
	Updated slapd/-lldap getaddrinfo handling (ITS#1304)
	Updated -lldap_r pthread stack size
	Updated slapd acl assert checks
	Updated slurpd ManageDSAIT handling
	Updated slapd root/subschema DSE update error handling
	Removed -lldbm use of Berkeley CDB
	Updated -lldbm to use r/w lock with DB_THREAD
	Updated back-ldbm update attribute indices on delete (ITS#1359)
	Updated ud Kerberos ticket handling
	Added back-ldbm idl_check() code
	Build environment
		Updated ltconfig for MacOS X 10.1 (ITS#1361)
		Updated slapd.conf with ACL example
		Updated pthread detection
		Updated Kerberos detection (ITS#1372)
		Updated test005 /dev/null diff for Tru64 compat
		Updated test007 modrdn test
		Updated test008 read/write ratio
		Updated MSVC5 project files
	Documentation
		Update slapcat/add pages

OpenLDAP 2.0.15 Release
	Fixed -lldap TLS external handling

servers/slapd/slappasswd.c
/* $OpenLDAP: pkg/ldap/servers/slapd/slappasswd.c,v 1.5.2.7 2010/04/13 20:23:21 kurt Exp $ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 1998-2011 The OpenLDAP Foundation.
 * Portions Copyright 1998-2003 Kurt D. Zeilenga.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* ACKNOWLEDGEMENTS:
 * This work was initially developed by Kurt Zeilenga for inclusion
 * in OpenLDAP Software.
 */

#include "portable.h"

#include <stdio.h>

#include <ac/stdlib.h>

#include <ac/ctype.h>
#include <ac/signal.h>
#include <ac/socket.h>
#include <ac/string.h>
#include <ac/time.h>
#include <ac/unistd.h>

#include <ldap.h>
#include <lber_pvt.h>
#include <lutil.h>
#include <lutil_sha1.h>

#include "ldap_defaults.h"
#include "slap.h"

static int	verbose = 0;

static void
usage(const char *s)
{
	fprintf(stderr,
		"Usage: %s [options]\n"
		"  -c format\tcrypt(3) salt format\n"
		"  -g\t\tgenerate random password\n"
		"  -h hash\tpassword scheme\n"
		"  -n\t\tomit trailing newline\n"
		"  -s secret\tnew password\n"
		"  -u\t\tgenerate RFC2307 values (default)\n"
		"  -v\t\tincrease verbosity\n"
		"  -T file\tread file for new password\n"
		, s );

	exit( EXIT_FAILURE );
}

int
slappasswd( int argc, char *argv[] )
{
#ifdef LUTIL_SHA1_BYTES
	char	*default_scheme = "{SSHA}";
#else
	char	*default_scheme = "{SMD5}";
#endif
	char	*scheme = default_scheme;

	char	*newpw = NULL;
	char	*pwfile = NULL;
	const char *text;
	const char *progname = "slappasswd";

	int		i;
	char		*newline = "\n";
	struct berval passwd = BER_BVNULL;
	struct berval hash;

	while( (i = getopt( argc, argv,
		"c:d:gh:ns:T:vu" )) != EOF )
	{
		switch (i) {
		case 'c':	/* crypt salt format */
			scheme = "{CRYPT}";
			lutil_salt_format( optarg );
			break;

		case 'g':	/* new password (generate) */
			if ( pwfile != NULL ) {
				fprintf( stderr, "Option -g incompatible with -T\n" );
				return EXIT_FAILURE;

			} else if ( newpw != NULL ) {
				fprintf( stderr, "New password already provided\n" );
				return EXIT_FAILURE;

			} else if ( lutil_passwd_generate( &passwd, 8 )) {
				fprintf( stderr, "Password generation failed\n" );
				return EXIT_FAILURE;
			}
			break;

		case 'h':	/* scheme */
			if ( scheme != default_scheme ) {
				fprintf( stderr, "Scheme already provided\n" );
				return EXIT_FAILURE;

			} else {
				scheme = ch_strdup( optarg );
			}
			break;

		case 'n':
			newline = "";
			break;

		case 's':	/* new password (secret) */
			if ( pwfile != NULL ) {
				fprintf( stderr, "Option -s incompatible with -T\n" );
				return EXIT_FAILURE;

			} else if ( newpw != NULL ) {
				fprintf( stderr, "New password already provided\n" );
				return EXIT_FAILURE;

			} else {
				char* p;
				newpw = ch_strdup( optarg );

				for( p = optarg; *p != '\0'; p++ ) {
					*p = '\0';
				}
			}
			break;

		case 'T':	/* password file */
			if ( pwfile != NULL ) {
				fprintf( stderr, "Password file already provided\n" );
				return EXIT_FAILURE;

			} else if ( newpw != NULL ) {
				fprintf( stderr, "Option -T incompatible with -s/-g\n" );
				return EXIT_FAILURE;

			}
			pwfile = optarg;
			break;

		case 'u':	/* RFC2307 userPassword */
			break;

		case 'v':	/* verbose */
			verbose++;
			break;

		default:
			usage ( progname );
		}
	}

	if( argc - optind != 0 ) {
		usage( progname );
	} 

	if( pwfile != NULL ) {
		if( lutil_get_filed_password( pwfile, &passwd )) {
			return EXIT_FAILURE;
		}
	} else if ( BER_BVISEMPTY( &passwd )) {
		if( newpw == NULL ) {
			/* prompt for new password */
			char *cknewpw;
			newpw = ch_strdup(getpassphrase("New password: "));
			cknewpw = getpassphrase("Re-enter new password: ");
	
			if( strcmp( newpw, cknewpw )) {
				fprintf( stderr, "Password values do not match\n" );
				return EXIT_FAILURE;
			}
		}

		passwd.bv_val = newpw;
		passwd.bv_len = strlen(passwd.bv_val);
	} else {
		hash = passwd;
		goto print_pw;
	}

	lutil_passwd_hash( &passwd, scheme, &hash, &text );
	if( hash.bv_val == NULL ) {
		fprintf( stderr,
			"Password generation failed for scheme %s: %s\n",
			scheme, text ? text : "" );
		return EXIT_FAILURE;
	}

	if( lutil_passwd( &hash, &passwd, NULL, &text ) ) {
		fprintf( stderr, "Password verification failed. %s\n",
			text ? text : "" );
		return EXIT_FAILURE;
	}

print_pw:;
	printf( "%s%s" , hash.bv_val, newline );
	return EXIT_SUCCESS;
}
